dump feature flags to user properties

mike-dydx opened the pull request:
## Description / Intuition
- replaces Amplitude-iOS ([deprecated](https://amplitude.com/docs/sdks/analytics/ios)) with Amplitude-Swift
- adds `statsigFlags` user property
- replaces `userInfo` with `set(userProperty:value:)` API
- dumps features flags to user properties after statsig initializes for the session
- firebase user property values are max 36 characters so we're adding a statsigUserId  user property to all events so we can look up a user's feature flag state for firebase. The Statsig-generated stable id was 37 characters lmao so had to add a custom stable ID generator and save it to user defaults
- moved amplitude injection earlier in the app lifecycle (otherwise user properties were getting set before amplitude was initialzied
- noticed that I introduced [this issue](https://github.com/firebase/firebase-ios-sdk/issues/12618#issuecomment-2016507842) when I created dydxAnalytics. I noticed it because Firebase DebugView was not working. Was seeing the same logs as the issue's description where Firebase libs were detected twice in DerivedData by Xcode. Essentially, we cannot add dydxAnalytics as a dependency of any other targets that are not dydxV4. Eventually, I might unwind dydxAnalytics and effectively put it into dydxV4 instead to avoid this mistake in the future.

<br/>

## Before/After Screenshots or Videos

| Before | After |
|--------|-------|
| fresh launch | <img src="https://github.com/user-attachments/assets/b91ebf47-ca64-4a49-b25e-936efb2464ed"> |
| 2nd launch | <img src="https://github.com/user-attachments/assets/da62ecb2-826b-4313-bd40-e004a8e2bddd"> |
| 3rd launch (after connect wallet on 2nd launch) | <img src="https://github.com/user-attachments/assets/e6a60806-0b99-4011-a659-5c31260d856e"> |
| amplitude user properties | <img src="https://github.com/user-attachments/assets/f5ed4c97-1473-4402-ab4a-367680dae0a1"> |
| firebase user properties | <img src="https://github.com/user-attachments/assets/e4804901-4200-4753-a5e3-24d07b1d6cd2"> |)


<br/>

### Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring or Technical Debt
- [ ] Documentation update
- [ ] Other (please describe: )

mike-dydx committed:
make userInfo [String: Any]

mike-dydx committed:
pod update, adopt amplitude swift

mike-dydx committed:
add statsig init listener to dydxCompositeTracking

mike-dydx commented on AmplitudeInjections/AmplitudeInjections/_Configure/AmplitudeRunner.swift:
moved this to `AmplitudeTracking` since the new Amplitude-Swift SDK apis works a little diffferently around initialization

mike-dydx commented on Shared/CommonAppDelegate.swift:
```diff
@@ -114,8 +114,7 @@ open class CommonAppDelegate: ParticlesAppDelegate {
             apiKey = CredientialConfig.shared.credential(for: "amplitudeApiKey")
         }
         if let apiKey = apiKey, apiKey.isNotEmpty {
-            AmplitudeRunner.shared.apiKey = apiKey
-            add(tracking: dydxAmplitudeTracking())
```
this allowed me to delete `AmplitudeRunner` class

mike-dydx committed:
remove schemes, re-apply pod update

mike-dydx commented on AmplitudeInjections/AmplitudeInjections/_Tracking/AmplitudeTracking.swift:
```diff
@@ -5,30 +5,17 @@
-        didSet {
-            let userIdKey = "walletAddress"
-            if let userInfo = userInfo {
-                if let userId = userInfo[userIdKey] {
-                    Console.shared.log("Amplitude: User ID set to walletAddress: \(userId ?? "nil")")
-                    Amplitude.instance().setUserId(userId)
-                }
```
refactored to `set(userId:)`

mike-dydx commented on AmplitudeInjections/AmplitudeInjections/_Tracking/AmplitudeTracking.swift:
```diff
@@ -5,30 +5,17 @@
-                thinned.removeValue(forKey: userIdKey)
-                Amplitude.instance().setUserProperties(thinned)
-            } else {
-                Amplitude.instance().setUserProperties(nil)
```
refactored to `setUserProperty(value:userProperty:)`

mike-dydx commented on FirebaseStaticInjections/FirebaseStaticInjections.xcodeproj/project.pbxproj:
```diff
@@ -391,7 +384,6 @@
 				31B3E3C022763B19009E8FEF /* FirebaseStaticInjections.swift */,
 				314B5E3723DCCD6700139EB3 /* _Configure */,
 				314B5E3B23DCCD6700139EB3 /* _ErrorLogging */,
-				314B5E3923DCCD6700139EB3 /* _FeatureFlags */,
```
this was empty

mike-dydx committed:
clean up

ruixhuang commented on dydxV4/dydxV4/_Tracking/dydxCompositeTracking.swift:
```diff
@@ -12,23 +12,13 @@ import dydxStateManager
-    
-    func set(userProperty: UserProperty, toValue value: String?) {
-        self.setUserInfo(key: userProperty.rawValue, value: value)
+    func setUserProperty(_ value: Any?, forUserProperty userProperty: UserProperty) {
```
Nit: How about renaming it to `setValue(_ value: Any?, forUserProperty: UserProperty)`?  It seems more intuitive.

mike-dydx replied:
i will in a follow-up PR

mike-dydx replied:
addressed d8364e2f0a87cf8ea21b5fe932d6301453b2162b

mike-dydx committed:
fixes & address PR feedback

move up injectAmplitude in the lifecycle

remove firebase from dydxAnalytics

rename function

chop characters to 36 for firebase user property

mike-dydx commented on FirebaseStaticInjections/FirebaseStaticInjections/_Tracking/FirebaseTracking.swift:
```diff
@@ -54,12 +54,18 @@ public class FirebaseTracking: TransformerTracker {
+    override public func setValue(_ value: Any?, forUserProperty userProperty: String) {
+        Console.shared.log("analytics log | Firebase: User Property `\(userProperty)` set to: \(value ?? "nil")")
+        // firebase max supported length is 36, this is best effort
+        if let valueString = parseAnyToString(value) {
+            Analytics.setUserProperty(String(valueString.prefix(36)), forName: userProperty)
+        } else {
+            Analytics.setUserProperty(nil, forName: userProperty)
+        }
```
figured it was better than nothing if we can get the first 36 characters of the property value we are intending to save
